accepted/2023/sign-signature-provider-plugins.md

```diff
@@ -0,0 +1,524 @@
+
+### Interfaces package
+
+We will create a new .NET assembly that contains only public interfaces to be implemented by plugins.  Sign CLI will implement a new command for a plugin that loads and interacts with the plugin implementation entirely by interfaces defined in the interfaces assembly.  This approach will enable Sign CLI and plugins to rev their implementations without either having any extraneous compile-time or runtime dependencies.
+
+Proposed interfaces:
+
+```C#
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+
+namespace Sign.Plugins.SignatureProvider.Interfaces
+{
+    /// <summary>
+    /// This is the core plugin interface.  Plugins conforming with this specification must implement this interface.
+    /// </summary>
+    public interface IPlugin
+    {
+        Task InitializeAsync(
+            IReadOnlyDictionary<string, string?> arguments,
+            ServiceProvider serviceProvider,
+            CancellationToken cancellationToken);
+    }
+
+    /// <summary>
+    /// This is the core interface for signature provider plugins.  Plugins conforming with this specification must
+    /// implement this interface.
+    /// </summary>
+    /// <remarks>
+    /// The plugin host will try casting a <see cref="IPlugin" /> instance to this interface to determine if the
+    /// plugin is a signature provider.
+    /// </remarks>
+    public interface ISignatureProvider
+    {
+        Task<AsymmetricAlgorithm> GetAsymmetricAlgorithmAsync(CancellationToken cancellationToken);
+    }
+
+    /// <summary>
+    /// This is an optional interface for a signature provider plugin.
+    /// If the asymmetric key pair returned by <see cref="ISignatureProvider" /> has an associated X.509 certificate,
+    /// the plugin should also implement this interface.
+    /// <summary>
+    /// <remarks>
+    /// The plugin host will try casting a <see cref="ISignatureProvider" /> instance to this interface to determine
+    /// if the plugin is a certificate provider.  A plugin should not implement this interface unless it can provide
+    /// a certificate.  Do not implement this interface and throw <see cref="NotImplementedException" /> to indicate
+    /// non-support.
+    ///
+    /// Some signature formatters work with raw asymmetric key pairs without a certificate.  To support only these
+    /// signature formatters, a plugin need not implement this interface.
+    ///
+    /// Some signature formatters (like Authenticode) do require a certificate and will fail without a certificate.
+    /// To support these signature formatters, the plugin should implement this interface.  Signature formatters may
+    /// fail if they require a certificate but a signature provider plugin does not implement this interface.
+    /// </remarks>
+    public interface ICertificateProvider
```

Comment:
I'd imagine that it would be desirable for `ICertificateProvider` to be a special type of `ISignatureProvider` as well. The definition could be updated to `public interface ICertificateProvider : ISignatureProvider` such that those interfaces that only support `ISignatureProvider` can still use a certificate acquired from a `ICertificateProvider`?